crypto/evp/evp_lib.c

```diff
@@ -76,13 +76,15 @@ int EVP_CIPHER_param_to_asn1(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
             goto err;
 
         /* ... but, we should get a return size too! */
-        if (params[0].return_size != 0
+        if (OSSL_PARAM_modified(params)
+            && params[0].return_size != 0
             && (der = OPENSSL_malloc(params[0].return_size)) != NULL) {
             params[0].data = der;
             params[0].data_size = params[0].return_size;
-            params[0].return_size = 0;
+            OSSL_PARAM_set_unmodified(params);
```

Comment:
Style nit: I think the intention is clearer if you make the argument `&params[0]`

Reply:
Ugh.

Reply:
Really???  I'd prefer to change the previous lines to `*params`.

Reply:
well, when I first saw `OSSL_PARAM_set_unmodified(params)`, the *first* thing that came to my mind was "wait what?  Is that function modifying the whole params array?"

Unfortunately, this is a fallacy in C, it's unclear if a passed pointer of an array is meant to just affect the first element or the whole array.

Maybe I'm overthinking this and it's just a matter of remembering what the function does...

Reply:
It could modify the whole array if we want....

I think that's overkill but might be easier for users.

Reply:
I'm warming to the idea of `OSSL_PARAM_set_unmodified` resetting the whole params array.

Reply:
There's no need for this function, since you're already directly setting three of the fields.

Reply:
You do understand, I hope, that I have the same hiccup at `OSSL_PARAM_modified(param)`...

I'm not sure it's wise to have an API where the related functions behave so very differently.  `OSSL_PARAM_set_modified()` and `OSSL_PARAM_modified()` should work in tandem, that's simply less confusing API design.

Reply:
It doesn't seem like there is a win here.  Do it one way, complaints.  Do it the other, complaints.

I'll change the API to `OSSL_PARAM_set_all_unmodified()` which makes it clearer.

Reply:
Yeah well, this got to bigger proportions than I had anticipated for a simple style nit. I wish I had shut the hell up...

Reply:
We all wish that at times :)

I think the new call is an improvement nonetheless.

Reply:
Okie.  At the very least, it cleared some confusion :wink: